[GR-34179] Upgrade CE debug info feature to provide information about Java types for Windows/PECOFF

stooke opened the pull request:
This PR adds type information to the CodeView 4 debug information (See PR #2396).  It is related to issue #2986, but is targeted to Windows and Visual Studio.

This PR is based off (and highly dependant on) the work in PR #3046.

With this patch, memory can be examined from the Visual Studio or WinDbg command line and cast to Java types as laid out by the Graal compiler.  Static member variables can be referenced directly.

Caveats:
- Pointer types for instance members in the presence of isolates are not yet implemented, as there appears to be no way to express a pointer that is an offset from r14 in CodeView.  I will continue to research this, but at this time, this implementation works best if isolates are disabled.
- As there is no local variable information added by this PR, users must manually cast registers to pointers; for example (java_lang_String*)(rdx). 

I recommend @adinn as a first reviewer for this PR, as it is based on his debug type PR.

olpaw commented:
@adinn looks like github does not allow me to put you on this one as reviewer :disappointed:

adinn commented:
@olpaw No worry. I'll review it anyway :-)

adinn commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeRecord.java:
```diff
@@ -117,10 +159,307 @@ private static int alignPadded4(byte[] buffer, int originalpos) {
+            GraalError.shouldNotReachHere();
+            return 0;
+        }
+
```
These should not reach here's smell a bit odd -- as if you ought to have an intermediate abstract subclass below CVTypeRecord for the all the subclasses that have content and hence implement these methods. That woudl allow you to rely on the type system to ensure that these methods don't get called.

Is there a reason why you cannot structure the class hierarchy like that? or is it just a question of the difficulty you would face retrofitting it into the original design?

stooke replied:
It's unclear what changes would be required for aarch64 debug info, so I thought I'd catch this one here, just to ensure it's not assumed to work.  I'm happy to remove the test altogether, but I'd rather not restructure the code until/unless a Windows aarch64 port is underway.

adinn replied:
Ok, so this is just future proofing That's fine.

adinn commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeSectionBuilder.java:
```diff
@@ -0,0 +1,651 @@
+     * We now have a class record but must create a pointer record. * ptrRecord = addTypeRecord(new
+     * CVTypeRecord.CVTypePointerRecord(clsRecord.getSequenceNumber(),
+     * CVTypeRecord.CVTypePointerRecord.NORMAL_64)); } return ptrRecord.getSequenceNumber(); }
+     */
```
This comment really needs removing. If you really want to keep the code you can mark it as unused.

stooke replied:
Removed - it was mangled and I didn't catch it.  It was only required if there was such a thing as a pointer to a primitive non-array type in Java - which there isn't.

adinn reviewed (approved):
This looks ok to me modulo the 3 inline comments I made. Not that I can really comment on the intricacies of the PeCOFF records and their detailed layout. 

I believe this is going to need to be modified slightly when rebased over head in order to take account of the changes made to support inline info. Specifically PrimaryEntry.getSubranges() is still being called from CVLinerecordBuilder. That has been dropped in head in favour of calling  leafRangeIterator(). I don't  think there us anything else needs changing though.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVDebugInfo.java:
```diff
@@ -36,13 +38,34 @@
+    /* See AMD64ReservedRegisters.java. */
+    public static final byte rheapbase_x86 = (byte) 14;
+    public static final byte rthread_x86 = (byte) 15;
+
```
Done.

adinn commented:
> I believe this is going to need to be modified slightly when rebased over head in order to take account of the changes made to support inline info. Specifically PrimaryEntry.getSubranges() is still being called from CVLinerecordBuilder. That has been dropped in head in favour of calling leafRangeIterator(). I don't think there us anything else needs changing though.

@stooke The above tweak turns out to be unneeded. The upstream code was updated to use the correct iterator when the gdb inline methods patch went in.  Once this PR is rebased it all fits together without any conflict. So, it appears to be ready for Oracle review and merging.

@pejovica Could you please review this and help get it through the Oracle gate?

pejovica commented:
@adinn Sure, I should be able to start with that after next week.

adinn commented:
@pejovica This PR is still waiting to be reviewed. will you be able to do that soon or do we need to find someone else to review it?

olpaw reviewed (changes requested):
Hi @stooke 
Great to see progress on the Windows side of CE debuginfo generation!

Please consolidate the commits. E.g. combine subsequent style related commits into one (or even better, fold them into the commit that is the origin of the style issue). Also start commit messages with uppercase. A more thorough review from @pejovica will follow.

olpaw commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVDebugInfo.java:
```diff
@@ -53,4 +76,12 @@ public CVTypeSectionImpl getCVTypeSection() {
+    @SuppressWarnings("unused")
+    public byte getThreadRegister() {
+        return threadRegister;
+    }
```
Given those accessor methods the the javadoc comments on the fields above are redundant.

stooke replied:
Comments removed.

olpaw commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVSymbolSubrecord.java:
```diff
@@ -246,24 +246,118 @@ public String toString() {
+        }
+
+        @SuppressWarnings("unused")
+        CVSymbolGData32Record(CVDebugInfo cvDebugInfo, String name, int typeIndex, int offset, short segment) {
```
Why keep it if it's unused?

stooke replied:
It was there for future use by references that are not relative.  Removed.

olpaw commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVSymbolSubrecord.java:
```diff
@@ -359,6 +454,31 @@ public String toString() {
     }
 
+    @SuppressWarnings("unused")
+    public static final class CVSymbolUDTRecord extends CVSymbolSubrecord {
```
Do you have the unused subclasses for future improvements or are the leftovers from prev. experiments?

stooke replied:
They are there for future use.

olpaw commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVSymbolSubsectionBuilder.java:
```diff
@@ -75,6 +96,27 @@ private void build(ClassEntry classEntry) {
+                addToSymbolSubsection(new CVSymbolSubrecord.CVSymbolRegRel32Record(cvDebugInfo, externName, typeIndex, f.getOffset(), cvDebugInfo.getHeapbaseRegister()));
+            } else {
+                /* Offset from heap begin. */
+                addToSymbolSubsection(new CVSymbolSubrecord.CVSymbolGData32Record(cvDebugInfo, externName, HEAP_BEGIN_NAME, typeIndex, f.getOffset(), sectionId));
```
Same way you are using `SectionName.SVM_HEAP` above you could use `DwarfDebugInfo.HEAP_BEGIN_NAME` here.

stooke replied:
Done - it's more clear as you suggest.

olpaw commented:
- [ ] Please consolidate the commits. E.g. combine subsequent style related commits into one (or even better, fold them into the commit that is the origin of the style issue). Also start commit messages with uppercase.

olpaw commented:
>  I've started to address your concerns, and squashed many of the commits, changing the messages to start with uppercase. Please let me know if I should squash further.

Looks much better now. Thanks!

adinn commented:
@olpaw Sure, thanks for the update and no inconvenience or problem about @pejovica taking his time. I'm very glad to hear he is giving it a proper workout.

pejovica reviewed (changes requested):
@stooke Thanks for working on this!

Overall, the PR looks good, but there are a few more things to do. Apart from what is stated in the comments, my main concern is the design of `CVTypeSectionBuilder`. In particular, I think there is no need to replace forwarding records with complete records on the fly as it only makes things harder to reason about.

Instead, I suggest establishing a few invariants that we can easily check:
  1. Type records should only use forwarding records for cross-referencing.
  2. Symbol records should only use complete records for cross-referencing (e.g., for `S_UDT` records).

This would mean that `CVTypeSectionBuilder.buildClass` will still make a complete record, but while doing so it will only be able to reference forwarding records. As a result, there will be no need to track whether the construction of a complete record is in progress -- it will no longer matter as only forwarding records will be referenced, and we can make them as needed.

stooke commented:
@pejovica thank you for the extensive review!  I'll be responding to the issues you've raised over the next few days.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVSymbolSubrecord.java:
```diff
@@ -246,24 +246,113 @@ public String toString() {
+    }
+
+    @SuppressWarnings("unused")
+    public static class CVSymbolLData32Record extends CVSymbolGData32Record {
```
Done.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVSymbolSubsectionBuilder.java:
```diff
@@ -75,6 +95,27 @@ private void build(ClassEntry classEntry) {
+                addToSymbolSubsection(new CVSymbolSubrecord.CVSymbolRegRel32Record(cvDebugInfo, externName, typeIndex, f.getOffset(), cvDebugInfo.getHeapbaseRegister()));
+            } else {
+                /* Offset from heap begin. */
+                addToSymbolSubsection(new CVSymbolSubrecord.CVSymbolGData32Record(cvDebugInfo, externName, DwarfDebugInfo.HEAP_BEGIN_NAME, typeIndex, f.getOffset(), sectionId));
```
Done.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVSymbolSubrecord.java:
```diff
@@ -311,20 +400,21 @@ protected int computeContents(byte[] buffer, int initialPos) {
-            return String.format("S_GPROC32   name=%s/%s parent=%d debugstart=0x%x debugend=0x%x len=0x%x offset=0x%x type=0x%x flags=0x%x)", debuggerName, externalName, pparent, debugStart, debugEnd,
-                            proclen, offset, typeIndex, flags);
+            return String.format("S_GPROC32   name=%s/%s parent=%d debugstart=0x%x debugend=0x%x len=0x%x seg:offset=0x%x:0x%x type=0x%x flags=0x%x)", debuggerName, externalName, pparent, debugStart,
+                            debugEnd,
+                            proclen, segment, offset, typeIndex, flags);
```
Done.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVUtil.java:
```diff
@@ -63,6 +72,30 @@ static int putInt(int i, byte[] buffer, int initialPos) {
         return pos;
     }
 
+    static int putLfNumeric(long i, byte[] buffer, int initialPos) {
```
I've added comments to most of the Util methods.  At this time I haven't redone this function but will probably address that later.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeSectionBuilder.java:
```diff
@@ -0,0 +1,633 @@
+    private final CVTypeSectionImpl typeSection;
+    private DebugContext debugContext;
+    private HeaderTypeEntry globalHeaderEntry;
+    private int depth = 0;
```
Removed.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVSymbolSubsectionBuilder.java:
```diff
@@ -75,6 +95,27 @@ private void build(ClassEntry classEntry) {
+            String externName = memberNameToCodeViewName(classEntry, f);
+            if (cvDebugInfo.useHeapBase()) {
+                /* REL32 offset from heap base register. */
+                addToSymbolSubsection(new CVSymbolSubrecord.CVSymbolRegRel32Record(cvDebugInfo, externName, typeIndex, f.getOffset(), cvDebugInfo.getHeapbaseRegister()));
```
I propose we definitely say we don't support debugging with isolates (should we emit a warning?), but I'd like to keep this code in; using pointer arithmetic (and a notepad) it is possible to make your way around even with isolates.

stooke replied:
Fixed the register constants properly.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVSymbolSubrecord.java:
```diff
@@ -246,24 +246,113 @@ public String toString() {
+        protected final int typeIndex;
+        protected final int offset;
+        protected final short segment;
+        protected final String name;
+        private final String relativeTo;
```
Done.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVSymbolSubrecord.java:
```diff
@@ -246,24 +246,113 @@ public String toString() {
+        private final int offset;
+        private final short segment;
+        private final byte flags;
+        private final String externalName;
+        private final String debuggerName;
```
Done.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVSymbolSubsectionBuilder.java:
```diff
@@ -75,6 +95,27 @@ private void build(ClassEntry classEntry) {
+        final short sectionId = (short) ((PECoffObjectFile.PECoffSection) rwSection).getSectionID();
+        classEntry.fields().filter(CVSymbolSubsectionBuilder::isManifestedStaticField).forEach(f -> {
+            int typeIndex = cvDebugInfo.getCVTypeSection().getIndexForPointer(f.getValueType());
+            String externName = memberNameToCodeViewName(classEntry, f);
```
Done.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeSectionImpl.java:
```diff
@@ -63,12 +83,14 @@ public String getSectionName() {
         pos = CVUtil.putInt(CV_SIGNATURE_C13, null, pos);
         for (CVTypeRecord record : typeMap.values()) {
             pos = record.computeFullSize(pos);
+            pos = CVUtil.pad4(null, pos);
```
I was getting a lot of messages from the codeview dump about misaligned records; this was part of an effort to fix it.  Turns out the linker will emit multiple class definitions and screw up alignment anyways, so probably this can be removed.

stooke replied:
removed.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeRecord.java:
```diff
@@ -235,4 +569,1064 @@ public boolean equals(Object obj) {
+        }
+    }
+
+    static final class CVTypeArrayRecord extends CVTypeRecord {
```
Done.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeRecord.java:
```diff
@@ -235,4 +569,1064 @@ public boolean equals(Object obj) {
+    }
+
+    @SuppressWarnings("unused")
+    static final class CVTypeServer2Record extends CVTypeRecord {
```
removed.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeSectionBuilder.java:
```diff
@@ -0,0 +1,633 @@
+        depth++;
+
+        /* Build return type records. */
+        /* TODO - build from proper type instead of name */
+        /* currently this builds ugly class records for arrays, so cheat by returning Object */
+        // TypeEntry foo = primary.getMethodEntry().getValueType();
+        // int returnTypeIndex =
+        // primary.getMethodEntry().getValueType().getTypeName().endsWith("[]") ?
+        // getIndexForPointer(JAVA_LANG_OBJECT) :
+        // getIndexForPointer(primary.getMethodEntry().getValueType().getTypeName());
```
removed.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeRecord.java:
```diff
@@ -235,4 +569,1064 @@ public boolean equals(Object obj) {
+        @Override
+        public int hashCode() {
+            int hash = type;
+            for (FieldRecord field : members) {
+                hash = 31 * hash + field.hashCode();
+            }
```
Easier to read, too.  (but a bit slower).  Changed.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeRecord.java:
```diff
@@ -235,4 +569,1064 @@ public boolean equals(Object obj) {
+                return false;
+            }
+            CVFieldListRecord other = (CVFieldListRecord) obj;
+            if (other.members.size() != members.size()) {
+                return false;
+            }
+            for (int i = 0; i < members.size(); i++) {
+                if (!members.get(i).equals(other.members.get(i))) {
+                    return false;
+                }
+            }
+            return true;
```
Done.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeRecord.java:
```diff
@@ -117,10 +159,307 @@ private static int alignPadded4(byte[] buffer, int originalpos) {
+    static final class CVTypeStringIdRecord extends CVTypeRecord {
+
+        String string;
+        int substringIdx;
```
Done.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeRecord.java:
```diff
@@ -117,10 +159,307 @@ private static int alignPadded4(byte[] buffer, int originalpos) {
+                return false;
+            }
+            CVTypeStringIdRecord other = (CVTypeStringIdRecord) obj;
+            return this.string.equals(other.string);
```
Done.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeRecord.java:
```diff
@@ -235,4 +569,1064 @@ public boolean equals(Object obj) {
+            h = 31 * h + funcAttr;
+            h = 31 * h + argList.getSequenceNumber();
+            h = 31 * h + thisType;
+            return h;
```
Done.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeRecord.java:
```diff
@@ -235,4 +569,1064 @@ public boolean equals(Object obj) {
+
+        @Override
+        public int hashCode() {
+            int h = type;
+            h = 31 * h + attrs;
```
Done.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeRecord.java:
```diff
@@ -235,4 +569,1064 @@ public boolean equals(Object obj) {
+                return false;
+            }
+            CVClassRecord other = (CVClassRecord) obj;
+            return this.count == other.count && this.propertyAttributes == other.propertyAttributes && this.fieldIndex == other.fieldIndex && this.derivedFromIndex == other.derivedFromIndex &&
```
Done.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeSectionBuilder.java:
```diff
@@ -0,0 +1,633 @@
+        if (debugContext != null) {
+            debugContext.logv(DebugContext.INFO_LEVEL, fmt, args);
+        } else {
+            System.out.format(fmt + "\n", args);
```
Done.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeRecord.java:
```diff
@@ -117,10 +159,307 @@ private static int alignPadded4(byte[] buffer, int originalpos) {
+        }
+
+        CVTypeStringIdRecord(String string) {
+            this(0, string);
```
Done.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeSectionBuilder.java:
```diff
@@ -0,0 +1,633 @@
+                    });
+                }
+                /* Only bother to build fieldlist record is there are actually manifested fields. */
+                if (fieldListRecord.count() > 0) {
```
Done.

stooke replied:
Done.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeRecord.java:
```diff
@@ -117,10 +159,307 @@ private static int alignPadded4(byte[] buffer, int originalpos) {
+    }
+
+    @SuppressWarnings("unused")
+    static final class CVUdtTypeLineModRecord extends CVUdtTypeLineRecord {
```
removed.

stooke replied:
Gone.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeSectionBuilder.java:
```diff
@@ -0,0 +1,633 @@
+    private CVTypeRecord.CVTypeMFunctionRecord buildMemberFunction(ClassEntry classEntry, MethodEntry methodEntry) {
+        CVTypeRecord.CVTypeMFunctionRecord mFunctionRecord = new CVTypeRecord.CVTypeMFunctionRecord();
+        mFunctionRecord.setClassType(getIndexForType(classEntry));
+        mFunctionRecord.setThisType(getIndexForPointerOrPrimitive(classEntry, false));
```
Done.

stooke replied:
Done.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeRecord.java:
```diff
@@ -235,4 +569,1064 @@ public boolean equals(Object obj) {
+        }
+    }
+
+    static final class CVMemberMethodRecord extends FieldRecord {
```
Done.

stooke replied:
Changed.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeRecord.java:
```diff
@@ -235,4 +569,1064 @@ public boolean equals(Object obj) {
+        }
+    }
+
+    static class CVBaseClassRecord extends CVTypeRecord {
```
Actually, it's more like a remnant from the future; it was added (a) for completeness, and (b) because I had a vague idea I might do something clever for interfaces later on. Removed.

stooke replied:
Gone.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeSectionBuilder.java:
```diff
@@ -0,0 +1,633 @@
+        CVTypeRecord.CVTypeMFunctionRecord mFunctionRecord = new CVTypeRecord.CVTypeMFunctionRecord();
+        mFunctionRecord.setClassType(getIndexForType(classEntry));
+        mFunctionRecord.setThisType(getIndexForPointerOrPrimitive(classEntry, false));
+        mFunctionRecord.setCallType((byte) (Modifier.isStatic(methodEntry.getModifiers()) ? CV_CALL_TYPE_C : CV_CALL_TYPE_THISCALL));
```
You're correct; it should always be CV_CALL_NEAR_C for x64.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeRecord.java:
```diff
@@ -235,4 +569,1064 @@ public boolean equals(Object obj) {
+             * TODO: there is some indication the offset is only present if attrs & (MPROP_VIRTUAL |
+             * MPROP_IVIRTUAL) != 0
+             */
+            if ((attrs & (MPROP_VIRTUAL | MPROP_IVIRTUAL)) != 0) {
```
Thanks - I missed this.  I've changed the handling of MPROP_IVIRTUAL accordingly.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeSectionBuilder.java:
```diff
@@ -0,0 +1,633 @@
+                            CVTypeRecord.CVTypeMFunctionRecord mFunctionRecord = buildMemberFunction(classEntry, m);
+                            addTypeRecord(mFunctionRecord);
+                            log("    overloaded method %s", mFunctionRecord);
+                            mlist.add((short) 0, (short) 0, mFunctionRecord.getSequenceNumber());
```
Done.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeRecord.java:
```diff
@@ -235,4 +569,1064 @@ public boolean equals(Object obj) {
+        private final int underlyingTypeIndex;
+
+        CVStaticMemberRecord(short attrs, int underlyingTypeIndex, String name) {
+            super(LF_STMEMBER, (short) (attrs + MPROP_STATIC), name);
```
Done.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeSectionBuilder.java:
```diff
@@ -0,0 +1,633 @@
+
+    private static short modifiersToAttr(MemberEntry member) {
+        short attr = Modifier.isPublic(member.getModifiers()) ? MPROP_PUBLIC : (Modifier.isPrivate(member.getModifiers())) ? MPROP_PRIVATE : MPROP_PROTECTED;
+        attr += Modifier.isStatic(member.getModifiers()) ? MPROP_STATIC : MPROP_VIRTUAL; // TODO_
```
I've tried to make this more intelligent based on the existence of a vtable slot.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeSectionBuilder.java:
```diff
@@ -0,0 +1,633 @@
+        mFunctionRecord.setThisType(getIndexForPointerOrPrimitive(classEntry, false));
+        mFunctionRecord.setCallType((byte) (Modifier.isStatic(methodEntry.getModifiers()) ? CV_CALL_TYPE_C : CV_CALL_TYPE_THISCALL));
+        short attr = modifiersToAttr(methodEntry);
+        mFunctionRecord.setFuncAttr((byte) attr);
```
I make the assumption that if name is "<init>" this is a ctor.

adinn replied:
I don't believe that is going to work. Method names passed to the debug info generator arrive via interface method MethodInfoProvider.name(). If you look at the [each](https://github.com/oracle/graal/blob/cad77e9b04319ad2c78318fe9d13acb2746179f8/substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/image/NativeImageDebugInfoProvider.java#L649) of the [three](https://github.com/oracle/graal/blob/cad77e9b04319ad2c78318fe9d13acb2746179f8/substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/image/NativeImageDebugInfoProvider.java#L870) available [implementations](https://github.com/oracle/graal/blob/cad77e9b04319ad2c78318fe9d13acb2746179f8/substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/image/NativeImageDebugInfoProvider.java#L1082) you will see that "<init>" gets translated to the simple name of the owning class

So, for example class org.my.Foo uses name `"Foo"`, inner class org.my.Foo.Bar uses `"Bar".` I think you need to look for cases where the method name matches the simple name.

stooke replied:
Thanks - I saw that "<clinit>" functions were coming through, so made an assumption without checking.  I've added a getSimpleName() method to ClassEntry and used it.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeRecord.java:
```diff
@@ -235,4 +569,1064 @@ public boolean equals(Object obj) {
+
+    static final class CVTypeMethodListRecord extends CVTypeRecord {
+
+        static class MDef {
```
I've made MDef inherit from CVOneMethodRecord, so a lot of the implementation is shared.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeSectionBuilder.java:
```diff
@@ -0,0 +1,633 @@
+        argListType = addTypeRecord(argListType);
+
+        /* Build actual type record */
+        CVTypeRecord funcType = addTypeRecord(new CVTypeRecord.CVTypeProcedureRecord().returnType(returnTypeIndex).argList(argListType));
```
Actually, I don't think this record needs to be generated at all, since the MFUNCTION is generated from the class definition.  Removed.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeSectionBuilder.java:
```diff
@@ -0,0 +1,633 @@
+    private CVTypeRecord.FieldRecord buildField(FieldEntry fieldEntry, boolean onlyForwardReference) {
+        TypeEntry valueType = fieldEntry.getValueType();
+        int vtIndex = getIndexForPointerOrPrimitive(valueType, onlyForwardReference);
+        short attr = modifiersToAttr(fieldEntry);
```
Done.  Are you sure static shouldn't be set?

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeSectionBuilder.java:
```diff
@@ -0,0 +1,633 @@
+        return record;
+    }
+
+    private CVTypeRecord buildStruct(StructureTypeEntry typeEntry, int superTypeIndex, String actualTypeName, CVTypeRecord.CVMemberRecord[] extraFields) {
```
done.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeSectionBuilder.java:
```diff
@@ -0,0 +1,633 @@
+        CVTypeRecord array0record = addTypeRecord(new CVTypeRecord.CVTypeArrayRecord(elementTypeIndex, T_UINT4, 0));
+
+        /* Build a field for the 0 length array. */
+        CVTypeRecord.CVMemberRecord dm = new CVTypeRecord.CVMemberRecord((short) 0x03, array0record.getSequenceNumber(), 0, "data");
```
done.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeSectionImpl.java:
```diff
@@ -107,14 +189,37 @@ private void addRecords() {
         final T record;
+        final boolean isInstance = newRecord.type == LF_CLASS || newRecord.type == LF_STRUCTURE;
+        /*
+         * Currently the hashes for identical class records do not always match. Until this is
+         * tracked down, use type name.
```
Not valid - removed.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeRecord.java:
```diff
@@ -235,4 +569,1064 @@ public boolean equals(Object obj) {
+        @Override
+        public int computeContents(byte[] buffer, int initialPos) {
+            int pos = CVUtil.putShort(count, buffer, initialPos);
+            pos = CVUtil.putShort((short) (propertyAttributes | ATTR_HAS_UNIQUENAME), buffer, pos);
```
Removed.  It is unclear to me if two strings get emitted if ATTR_HAS_UNIQUENAME is set.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVSymbolSubsectionBuilder.java:
```diff
@@ -75,6 +95,27 @@ private void build(ClassEntry classEntry) {
+    }
+
+    private static String memberNameToCodeViewName(ClassEntry classEntry, FieldEntry fieldEntry) {
+        return classEntry.getTypeName().replace(".", "_") + "::" + fieldEntry.fieldName();
```
I've added this (see the new class CVNames.java) but I'm looking for suggestions on a good algorithm. 
Currently you might see something like "java_util_Formatter_Flags_contains_java_util_Formatter_Flags" for java.util.Formatter.Flags.contains(Flags)

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVSymbolSubrecord.java:
```diff
@@ -297,11 +386,11 @@ protected int computeContents(byte[] buffer, int initialPos) {
             pos = CVUtil.putInt(typeIndex, buffer, pos);
             if (buffer != null) {
-                cvDebugInfo.getCVSymbolSection().markRelocationSite(pos, ObjectFile.RelocationKind.SECREL_4, externalName, false, 1L);
+                cvDebugInfo.getCVSymbolSection().markRelocationSite(pos, ObjectFile.RelocationKind.SECREL_4, externalName, false, 0L);
```
I've just make a single markRelocationSite() function, since everywhere requires both a section and offset.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVSymbolSubsectionBuilder.java:
```diff
@@ -75,6 +95,27 @@ private void build(ClassEntry classEntry) {
             build(primaryEntry);
         }
+        /* Add manifested static fields as S_GDATA32 records. */
+        final short sectionId = (short) ((PECoffObjectFile.PECoffSection) rwSection).getSectionID();
```
removed.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeRecord.java:
```diff
@@ -235,4 +569,1064 @@ public boolean equals(Object obj) {
+            pos = CVUtil.putInt(derivedFromIndex, buffer, pos);
+            pos = CVUtil.putInt(vshapeIndex, buffer, pos);
+            pos = CVUtil.putLfNumeric(size, buffer, pos);
+            String fixedName = className.replace('.', '_').replace("[]", "_array").replace('$', '_');
```
Helper class CVNames created, and '$' is stripped no more.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeRecord.java:
```diff
@@ -235,4 +569,1064 @@ public boolean equals(Object obj) {
+        }
+    }
+
+    static final class CVTypeMethodListRecord extends CVTypeRecord {
```
Although I've cleaned up and unified some of this code, I haven't done this at this time.  I don't think the savings are really worth it, but I'm open if you'd really like to see this.

stooke commented:
@pejovica , thanks again for the review.  I'm made extensive changes in line with your suggested improvements.  Some conversations I left unresolved as I might have a question or two.  At this point, I'd like to see if there are further changes you'd like to see before I merge the latest Graal head into this branch.

pejovica commented:
@stooke You can update right away as I'd be doing testing on the latest master anyway.

And I'll do another review round in the next few days.

stooke commented:
@pejovica, I've updated, but (like many other PRs) I can't pass the debugInfo gates because of issue #4018. 
Please let me know if there's anything I can address in the meantime.

pejovica commented:
@stooke Sorry for the delay. There were a few unrelated issues that I had to take care of, so unfortunately I didn't have enough time to do a proper review. I should be able to continue reviewing this PR next week.

adinn commented:
@pejovica This issue is still waiting for a review after from before Christmas. I believ the debuginfotets gate problem is no longer stopping it progressing. Are you likely to get time to review it soon?

pejovica commented:
@adinn Yes, and I've actually already started, so I hope to finish in the next few days.

pejovica commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/debugentry/ClassEntry.java:
```diff
@@ -369,4 +374,9 @@ public MethodEntry ensureMethodEntryForDebugRangeInfo(DebugRangeInfo debugRangeI
         return methods;
     }
+
+    public String getSimpleClassName() {
```
This seems to be necessary solely to determine whether a method is a constructor. I think it would be cleaner to introduce an API than to rely on a naming convention, say `DebugMethodInfo.isConstructor`, which could also be done as part of #4333.

stooke replied:
Done.

pejovica reviewed (commented):
@stooke Here are a few more bits that could be done as part of #4333.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVSymbolSubrecord.java:
```diff
@@ -75,7 +74,7 @@ public int getCommand() {
     public static final class CVObjectNameRecord extends CVSymbolSubrecord {
 
-        String objName; /* find the full path to object file we will produce. */
+        private final String objName; /* find the full path to object file we will produce. */
```
done.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/debugentry/ClassEntry.java:
```diff
@@ -119,6 +120,10 @@ public ClassEntry(String className, FileEntry fileEntry, int size) {
         }
     }
 
+    public Stream<MethodEntry> methods() {
```
Already gone.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVUtil.java:
```diff
@@ -28,55 +28,125 @@
+     * @param initialPos initial position in buffer
+     * @return position in buffer following stored value
+     */
+    static int putLfNumeric(long value, byte[] buffer, int initialPos) {
```
Done.

stooke commented:
Hello @pejovica , and thanks for the review - I have incorporated the changes suggested into the prereq PR #4333, and will update this PR shortly.

pejovica commented:
@stooke Excellent, thanks!

stooke commented:
@pejovica , I have updated this PR to reflect the changes from the prereq PR.  
While doing so, I've simplified CVTypeSectionImpl, and refactored CVTypeSectionBuilder to separate out utility functions (getIndexForPointerorPrimitive() for example) into a utility class.

pejovica commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVSymbolSubrecord.java:
```diff
@@ -295,7 +378,7 @@ protected int computeContents(byte[] buffer, int initialPos) {
             pos = CVUtil.putInt(debugEnd, buffer, pos);
             pos = CVUtil.putInt(typeIndex, buffer, pos);
-            pos = cvDebugInfo.getCVSymbolSection().markRelocationSite(buffer, pos, symbolName);
+            pos = cvDebugInfo.getCVSymbolSection().markRelocationSite(buffer, pos, symbolName, offset);
```
Not sure what the intention was here, but adding a non-zero addend to a relocation that already calculates the correct value is wrong. Please revert this change.

stooke replied:
I have removed the offset parameter from this code.  To me, it should stay, because this is a generic implementation of a S_GPROC32 record.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVSymbolSubrecord.java:
```diff
@@ -245,6 +245,89 @@ public String toString() {
+        @Override
+        protected int computeContents(byte[] buffer, int initialPos) {
+            int pos = CVUtil.putInt(typeIndex, buffer, initialPos);
+            String extname = symbolName != null ? symbolName : displayName;
```
Fixed.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVSymbolSubsectionBuilder.java:
```diff
@@ -51,27 +55,66 @@ final class CVSymbolSubsectionBuilder {
+        int classIndex = addTypeRecords(classEntry);
+
+        /*
+         * Adding an S_UDT (User Defined Type) record ensures the linker doesn't throw away the
+         * class definition.
```
Done.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVSymbolSubsectionBuilder.java:
```diff
@@ -116,27 +162,21 @@ private void build(PrimaryEntry primaryEntry) {
-            /* In the future, use a more user-friendly name instead of a hash function. */
-            methodName = range.getSymbolName();
+            /* Use a more user-friendly name instead of a hash function. */
+            methodName = CVNames.functionNameAndArgsToCodeViewName(range.getMethodEntry());
```
The problem comes when users try to type overloaded function names.  It's true that I could probably only add arguments to overloaded methods, but I suspect that is a more inconsistent UX.
Either way, I feel this should be addressed in the followup PR, which addresses symbol definitions more than type definitions.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVSymbolSubsectionBuilder.java:
```diff
@@ -51,27 +55,66 @@ final class CVSymbolSubsectionBuilder {
+        classEntry.fields().filter(CVSymbolSubsectionBuilder::isManifestedStaticField).forEach(f -> {
+            int typeIndex = cvDebugInfo.getCVTypeSection().getIndexForPointer(f.getValueType());
+            String displayName = CVNames.fieldNameToCodeViewName(f);
+            if (cvDebugInfo.useHeapBase()) {
```
Graal issue #4617 has been created.
This code actually works in the presence of isolates (unless certain falg bits are set), so I don't think it should be removed.

adinn commented:
@fniephaus @pejovica @olpaw It is getting very close to cutoff for the next release and we would very much like to see this patch go in. Simon believe he has provided a suitable response to all outstanding review requests. Can we push it as is, leaving any further pending changes to follow-up PRs or, failing that, can we work quickly to resolve any outstanding problems so we can make the release?

pejovica reviewed (commented):
@adinn There are still a few problems with the correctness that need to be resolved before merging.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeSectionBuilder.java:
```diff
@@ -0,0 +1,484 @@
+    private static final String JAVA_LANG_OBJECT_NAME = "java.lang.Object";
+    private static final String OBJ_HEADER_NAME = "_objhdr";
+
+    private static final int FUNC_IS_CONSTRUCTOR = 2;
```
Done.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeSectionBuilder.java:
```diff
@@ -0,0 +1,484 @@
+                log("building fields %s", fieldListRecord);
+
+                if (superTypeIndex != 0) {
+                    CVTypeRecord.CVBaseMemberRecord btype = new CVTypeRecord.CVBaseMemberRecord((short) 0x3, superTypeIndex, 0);
```
Done.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeSectionBuilder.java:
```diff
@@ -0,0 +1,484 @@
+                    CVTypeRecord array0record = addTypeRecord(new CVTypeRecord.CVTypeArrayRecord(elementTypeIndex, T_UINT4, 0));
+
+                    /* Build a field for the 0 length array. */
+                    CVTypeRecord.CVMemberRecord dm = new CVTypeRecord.CVMemberRecord((short) 0x03, array0record.getSequenceNumber(), typeEntry.getSize(), "data");
```
Done.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeSectionBuilder.java:
```diff
@@ -0,0 +1,484 @@
+    }
+
+    private void log(String fmt, Object... args) {
+        if (typeSection.getDebugContext() != null) {
+            typeSection.log(fmt, args);
+        }
```
Done.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeSectionBuilder.java:
```diff
@@ -0,0 +1,484 @@
+                 */
+                int line = classEntry.getPrimaryEntries().isEmpty() ? 0 : classEntry.getPrimaryEntries().get(0).getPrimary().getLine();
+                if (line > 0) {
+                    int idIdx = typeSection.getStringId(classEntry.getFileName()).getSequenceNumber();
```
Done.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeRecord.java:
```diff
@@ -235,4 +407,741 @@ public boolean equals(Object obj) {
+            pos = CVUtil.putShort(attrs, buffer, pos);
+            pos = CVUtil.putInt(funcIdx, buffer, pos);
+            if (hasVtableOffset()) {
+                pos = CVUtil.putInt(vtbleOffset, buffer, pos);
```
Done.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeSectionBuilder.java:
```diff
@@ -0,0 +1,484 @@
+        /* TODO take abstract (= pure) and vtableOffset into account */
+        if (isStatic) {
+            attr += MPROP_STATIC;
+        } else if (member.getVtableOffset() < 0) {
```
Done.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeSectionBuilder.java:
```diff
@@ -0,0 +1,484 @@
+    }
+
+    private static short modifiersToAttr(MethodEntry member) {
+        short attr = Modifier.isPublic(member.getModifiers()) ? MPROP_PUBLIC : (Modifier.isPrivate(member.getModifiers())) ? MPROP_PRIVATE : MPROP_PROTECTED;
```
Done.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeSectionBuilder.java:
```diff
@@ -0,0 +1,484 @@
+
+    private static short modifiersToAttr(FieldEntry member) {
+        short attr = Modifier.isPublic(member.getModifiers()) ? MPROP_PUBLIC : (Modifier.isPrivate(member.getModifiers())) ? MPROP_PRIVATE : MPROP_PROTECTED;
+        attr += Modifier.isStatic(member.getModifiers()) ? MPROP_STATIC : 0;
```
Done.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeSectionBuilder.java:
```diff
@@ -0,0 +1,484 @@
+                            addTypeRecord(mFunctionRecord);
+                            short attr = modifiersToAttr(m);
+                            log("    overloaded method %s", mFunctionRecord);
+                            int vtbleOffset = m.getVtableOffset() >= 0 && ((attr & MPROP_VSF_MASK) == MPROP_IVIRTUAL || (attr & MPROP_VSF_MASK) == MPROP_PURE_IVIRTUAL) ? m.getVtableOffset() : 0;
+                            mlist.add(attr, mFunctionRecord.getSequenceNumber(), vtbleOffset, m.methodName());
```
When I dump VC++ records, I only see the vtable offset on *IVIRTUAL records, otherwise set to 0.
It is strange, I know.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeRecord.java:
```diff
@@ -235,4 +407,741 @@ public boolean equals(Object obj) {
+        }
+
+        boolean hasVtableOffset() {
+            return vtbleOffset >= 0 && ((attrs & MPROP_VSF_MASK) == MPROP_IVIRTUAL || (attrs & MPROP_VSF_MASK) == MPROP_PURE_IVIRTUAL);
```
Done.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeSectionBuilder.java:
```diff
@@ -0,0 +1,484 @@
+    private CVTypeRecord buildStructureTypeEntry(final StructureTypeEntry typeEntry) {
+
+        CVTypeRecord typeRecord = types.getExistingType(typeEntry);
+        if (typeRecord == null || typeRecord.type == LF_CLASS && ((CVTypeRecord.CVClassRecord) typeRecord).isForwardRef()) {
```
Done.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeSectionBuilder.java:
```diff
@@ -0,0 +1,484 @@
+        if (typeRecord == null || typeRecord.type == LF_CLASS && ((CVTypeRecord.CVClassRecord) typeRecord).isForwardRef()) {
+            log("classentry size=%d kind=%s %s", typeEntry.getSize(), typeEntry.typeKind().name(), typeEntry.getTypeName());
+
+            ClassEntry superType = typeEntry.isClass() ? ((ClassEntry) typeEntry).getSuperClass() : null;
+            int superTypeIndex = superType != null ? types.getIndexForForwardRef(superType) : 0;
```
Done.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeSectionBuilder.java:
```diff
@@ -0,0 +1,484 @@
+            ClassEntry superType = typeEntry.isClass() ? ((ClassEntry) typeEntry).getSuperClass() : null;
+            int superTypeIndex = superType != null ? types.getIndexForForwardRef(superType) : 0;
+
+            /* Both java.lang.Object and __objhdr have null superclass. */
+            /* Force java.lang.Object to have __objhdr as a superclass. */
```
Done.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeSectionBuilder.java:
```diff
@@ -0,0 +1,484 @@
+                        methods.stream().filter(methodEntry -> methodEntry.methodName().equals(mname)).forEach(m -> {
+                            log("overloaded method %s(%s) attr=(%s) valuetype=%s", m.fieldName(), m.methodName(), m.getModifiersString(), m.getValueType().getTypeName());
+                            CVTypeRecord.CVTypeMFunctionRecord mFunctionRecord = buildMemberFunction((ClassEntry) typeEntry, m);
+                            addTypeRecord(mFunctionRecord);
```
Done.

stooke commented:
@pejovica , thanks for this review cycle, and especially the catch of the reversal of typeindex and vtable in one of the type records.  As far as I know, I have addressed this latest round.

adinn commented:
@fniephaus @pejovica @olpaw We would very much like to see this patch go in. @stooke believes he has provided a suitable response to all outstanding review requests. Can we push it as is, leaving any further pending changes to follow-up PRs or, failing that, can we work quickly to resolve any outstanding problems?

fniephaus commented:
Ok so it seems most gates are passing except for unfortunately some that build the EE version of `gu.exe`:
```
Fatal error: java.lang.RuntimeException: There was an error linking the native image: Linker command exited with 2
...
gu.obj : fatal error LNK1103: debugging information corrupt; recompile module
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.image.NativeImageViaCC.handleLinkerFailure(NativeImageViaCC.java:188)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.image.NativeImageViaCC.write(NativeImageViaCC.java:135)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGenerator.doRun(NativeImageGenerator.java:702)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGenerator.run(NativeImageGenerator.java:519)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGeneratorRunner.buildImage(NativeImageGeneratorRunner.java:407)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGeneratorRunner.build(NativeImageGeneratorRunner.java:585)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGeneratorRunner.main(NativeImageGeneratorRunner.java:128)
```

I'm not sure how we could provide a reproducer but maybe @pejovica or @olpaw have an idea how to fix this.

stooke commented:
> Ok so it seems most gates are passing except for unfortunately some that build the EE version of `gu.exe`:
> 
> ```
> Fatal error: java.lang.RuntimeException: There was an error linking the native image: Linker command exited with 2
> ...
> gu.obj : fatal error LNK1103: debugging information corrupt; recompile module
> 	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.image.NativeImageViaCC.handleLinkerFailure(NativeImageViaCC.java:188)
> 	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.image.NativeImageViaCC.write(NativeImageViaCC.java:135)
> 	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGenerator.doRun(NativeImageGenerator.java:702)
> 	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGenerator.run(NativeImageGenerator.java:519)
> 	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGeneratorRunner.buildImage(NativeImageGeneratorRunner.java:407)
> 	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGeneratorRunner.build(NativeImageGeneratorRunner.java:585)
> 	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGeneratorRunner.main(NativeImageGeneratorRunner.java:128)
> ```
> 
> I'm not sure how we could provide a reproducer but maybe @pejovica or @olpaw have an idea how to fix this.

If I can get hold of the object file, I might be able to see what's out of place.

stooke commented:
> > If I can get hold of the object file, I might be able to see what's out of place.
> 
> Sounds good. Maybe we can provide that.
> 
> Anyway, I just realized that we don't use the the CE debuginfo in EE yet, which may be causing this issue. I'll check with @olpaw next week.

@fniephaus , any word on this?  As this is an issue that shows up only in EE, and I don't have access to the EE code, and the EE version doesn't support Windows debugging (as I understand; please correct me if I'm mistaken), I suggest moving forward with this at this time.

fniephaus commented:
> any word on this?

As you may know, we're in a feature freeze atm and the team is busy preparing the upcoming release. Unfortunately, creating a reproducer for you turned out harder than expected, which is why we had to deprioritize this. I hope we can follow up very soon.

pejovica commented:
> > Ok so it seems most gates are passing except for unfortunately some that build the EE version of `gu.exe`:
> > ```
> > Fatal error: java.lang.RuntimeException: There was an error linking the native image: Linker command exited with 2
> > ...
> > gu.obj : fatal error LNK1103: debugging information corrupt; recompile module
> > 	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.image.NativeImageViaCC.handleLinkerFailure(NativeImageViaCC.java:188)
> > 	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.image.NativeImageViaCC.write(NativeImageViaCC.java:135)
> > 	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGenerator.doRun(NativeImageGenerator.java:702)
> > 	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGenerator.run(NativeImageGenerator.java:519)
> > 	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGeneratorRunner.buildImage(NativeImageGeneratorRunner.java:407)
> > 	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGeneratorRunner.build(NativeImageGeneratorRunner.java:585)
> > 	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGeneratorRunner.main(NativeImageGeneratorRunner.java:128)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I'm not sure how we could provide a reproducer but maybe @pejovica or @olpaw have an idea how to fix this.
> 
> If I can get hold of the object file, I might be able to see what's out of place.

@stooke I think I got it. We can generate `LF_FIELDLIST` records that are larger than 64 KB, and since we only have 16 bits for the length, we end up with records that can no longer be parsed.

So it looks like we'll need to add support for splitting large field lists into two or more records using `LF_INDEX` subrecords.

stooke commented:
@pejovica thanks for your investigation.  I have investigated the format of LF_INDEX and implemented it for LF_FIELDLIST.  At the same time, I added a check for oversize typeinfo records.

fniephaus commented:
@stooke thanks! Let me know when we can run the internal gate again. Should we squash some of the commits in this PR?

stooke commented:
> Good news: looks like the fix is working and the internal gates are passing now. Let me know when this is squashed and ready for merging.

@fniephaus I have rebased and squashed this PR.  It should be ready for merging.

fniephaus commented:
Feel free to ping me here or on Slack if any public gates need to be restarted. If you add new commits, I need to sync them with the internal PR and that means I need to re-run all internal gates, too.

stooke commented:
> Feel free to ping me here or on Slack if any public gates need to be restarted. If you add new commits, I need to sync them with the internal PR and that means I need to re-run all internal gates, too.

Sorry, I just added a dummy commit to restart the public gates; should I squash that and push again? Next time I'll ping you instead.

fniephaus commented:
It's all good... just wanted you to be aware that I need to sync all commits and that there are other ways to restart gates. :)

stooke commented:
> It's all good... just wanted you to be aware that I need to sync all commits and that there are other ways to restart gates. :)

Duly noted. :) Should I squash my dummy commit?  The PR is good ether way.

fniephaus commented:
Good! :) No need to do anything, I'll just drop the dummy commit so I don't need to restart the internal gates.

pejovica commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeRecord.java:
```diff
@@ -69,17 +100,23 @@ int computeFullSize(int initialPos) {
-        int length = (short) (pos - initialPos - Short.BYTES);
+        int length = pos - initialPos - Short.BYTES;
+        if (length > CV_TYPE_RECORD_MAX_SIZE) {
+            GraalError.shouldNotReachHere(String.format("Type record too large: %d (maximum %d) bytes: %s", length, CV_TYPE_RECORD_MAX_SIZE, this));
```
We usually use `throw` in cases like this.
```suggestion
            throw GraalError.shouldNotReachHere(String.format("Type record too large: %d (maximum %d) bytes: %s", length, CV_TYPE_RECORD_MAX_SIZE, this));
```

pejovica commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeRecord.java:
```diff
@@ -235,4 +413,806 @@ public boolean equals(Object obj) {
+
+        @Override
+        public boolean equals(Object obj) {
+            if (!super.equals(obj)) {
```
While investigating the issue with the `LF_FIELDLIST` records, I noticed that this actually calls `Object.equals`, which is probably not what you intended.

stooke replied:
Nope, thanks for the catch.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeRecord.java:
```diff
@@ -235,4 +413,806 @@ public boolean equals(Object obj) {
+            /* Keep a running total. */
+            estimatedSize += m.computeSize();
+            /* Pad to next 4 byte boundary if required. */
+            while ((estimatedSize % 4) != 0) {
+                estimatedSize++;
+            }
```
Actually, to simplify further:
estimatedSize += CVUtil.align4(m.computeSize());

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeSectionBuilder.java:
```diff
@@ -0,0 +1,538 @@
+        CVTypeRecord.CVFieldListRecord currentFieldList;
+        final Deque<CVTypeRecord.CVFieldListRecord> lists = new LinkedList<>();
+
+        FieldListContext(CVTypeRecord.CVFieldListRecord initialRecord) {
```
Refactored and renamed.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeSectionBuilder.java:
```diff
@@ -0,0 +1,538 @@
+        /* Build fieldlist record from manifested fields. */
+        CVTypeRecord.CVFieldListRecord newfieldListRecord = fieldListContext.addTypeRecords(this);
+        int fieldListIdx = newfieldListRecord.getSequenceNumber();
+        int fieldListCount = newfieldListRecord.count();
```
Thank you for catching this.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeRecord.java:
```diff
@@ -235,4 +413,806 @@ public boolean equals(Object obj) {
+        private final int index; /* index of continuation record */
+
+        CVIndexRecord(int index) {
+            /* Stick 'count' into the attrs field just for convenience. */
```
Done.  Actually I improved readability by making 'count' a member of the class instead of reusing attr.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeSectionBuilder.java:
```diff
@@ -0,0 +1,538 @@
+    }
+
+    static class FieldListContext {
+        /* Knock 100 bytes off to be safe from off-by-1, padding, LF_INDEX, etc. */
+        static final int FIELD_LIST_MAX_SIZE = CV_TYPE_RECORD_MAX_SIZE - 100;
```
You just know there's an off-by-one error waiting to happen.  I've removed this constant and assumed the math is correct.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeRecord.java:
```diff
@@ -235,4 +413,806 @@ public boolean equals(Object obj) {
+        }
+
+        int getEstimatedSize() {
+            /* Add in size of record type and length fields. */
+            return estimatedSize + Short.BYTES * 2;
```
@pejovica, I've refactored this to correct the calculation and alignment calculations, but I haven't moved to using 'availableSize' as I think the concept of limits is best kept to the FieldListBuilder.
If this is a showstopper, please let me know and we can discuss it.

stooke commented:
@pejovica, thank you again for your comments.  I have posted a set of changes to address them (but haven't adopted 'availableSize' for reasons given above).

pejovica reviewed (commented):
@stooke I left you a few more comments, but other than that the PR looks good.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVNames.java:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    static String staticFieldNameToCodeViewName(FieldEntry fieldEntry) {
+        return typeNameToCodeViewName(fieldEntry.ownerType()) + "_" + fieldEntry.fieldName();
```
Done.  Personally my taste is to keep the implementations separate but I've adopted this change.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeRecord.java:
```diff
@@ -117,56 +154,150 @@ private static int alignPadded4(byte[] buffer, int originalpos) {
-            return this;
+        @Override
+        public int computeSize(int initialPos) {
+            GraalError.shouldNotReachHere();
+            return 0;
```
Done.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeSectionBuilder.java:
```diff
@@ -0,0 +1,549 @@
+        }
+
+        void testForUndefinedClasses() {
+            /* Currently java.lang.Class should be the only type undefined at this point. */
```
Gone.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeSectionBuilder.java:
```diff
@@ -0,0 +1,549 @@
+            });
+        }
+        /* Build fieldlist record from manifested fields. */
+        CVTypeRecord.CVFieldListRecord newfieldListRecord = fieldListBuilder.buildFieldListRecords(this);
```
Done.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeRecord.java:
```diff
@@ -235,4 +413,790 @@ public boolean equals(Object obj) {
+
+        @Override
+        public String toString() {
+            String attrString = (funcAttr & 2) == 2 ? "(ctor)" : "";
```
Done.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeRecord.java:
```diff
@@ -235,4 +413,790 @@ public boolean equals(Object obj) {
+        }
+
+        @Override
+        public boolean equals(Object obj) {
```
Done.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeRecord.java:
```diff
@@ -235,4 +413,790 @@ public boolean equals(Object obj) {
+        private final int index; /* index of continuation record */
+
+        CVIndexRecord(int index) {
+            /* Stick 'count' into the attrs field just for convenience. */
```
Gone.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVSymbolSubsectionBuilder.java:
```diff
@@ -158,23 +185,32 @@ private void addLineNumberRecords(PrimaryEntry primaryEntry) {
+         * Adding an S_UDT (User Defined Type) record ensures the linker doesn't throw away the type
+         * definition.
+         */
+        CVSymbolSubrecord.CVSymbolUDTRecord udtRecord = new CVSymbolSubrecord.CVSymbolUDTRecord(cvDebugInfo, typeIdx, CVNames.typeNameToCodeViewName(typeEntry.getTypeName()));
```
Done.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVSymbolSubsectionBuilder.java:
```diff
@@ -116,29 +157,15 @@ private void build(PrimaryEntry primaryEntry) {
-            methodName = range.getSymbolName();
-        }
-        return methodName;
+    private static String getDebuggerName(Range range) {
```
Done.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeRecord.java:
```diff
@@ -235,4 +413,790 @@ public boolean equals(Object obj) {
+            pos = CVUtil.putLfNumeric(size, buffer, pos);
+            String fixedName = CVNames.typeNameToCodeViewName(className);
+            pos = CVUtil.putUTF8StringBytes(fixedName, buffer, pos);
+            if (uniqueName != null) {
+                pos = CVUtil.putUTF8StringBytes(uniqueName, buffer, pos);
+            } else {
+                pos = CVUtil.putUTF8StringBytes(fixedName, buffer, pos);
+            }
```
Done.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeRecord.java:
```diff
@@ -235,4 +413,790 @@ public boolean equals(Object obj) {
+            this.vshapeIndex = vshapeIndex;
+            this.size = size;
+            this.className = className;
+            this.uniqueName = null;
```
Done.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeSectionBuilder.java:
```diff
@@ -0,0 +1,549 @@
+
+            /* Build all Field List records in field order (FIFO). */
+            for (CVTypeRecord.FieldRecord fieldRecord : fields) {
+                if (currentFieldList.getEstimatedSize() + CVUtil.align4(fieldRecord.computeSize()) >= CV_TYPE_RECORD_MAX_SIZE) {
```
My original approach was to simply set the maximum size to (true_maximum_size - margin_for_error) which was both sloppy and guaranteed to work.
I've added FIELD_LIST_RECORD_OVERHEAD to account for the extra.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeRecord.java:
```diff
@@ -235,4 +413,790 @@ public boolean equals(Object obj) {
+
+        @Override
+        protected int computeContents(byte[] buffer, int initialPos) {
+            int pos = CVTypeRecord.alignPadded4(buffer, initialPos);
```
Actually, no.  It's enforced that it's already aligned on a four byte boundary at this point.  I'll take this code out as it may not need to be, but I'll leave the second alignPadded4() in so the individual fields will stay aligned.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeSectionBuilder.java:
```diff
@@ -0,0 +1,549 @@
+        objectHeaderRecordIndex = types.getIndexForForwardRef(OBJ_HEADER_NAME);
+    }
+
+    void buildRemainingRecords() {
```
Done.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeSectionImpl.java:
```diff
@@ -97,6 +106,7 @@ public void writeContent(DebugContext debugContext) {
     private void addRecords() {
         /* if an external PDB file is generated, add CVTypeServer2Record */
         /* for each class, add all members, types, etc */
+        builder.buildRemainingRecords();
```
I deleted this method.

stooke commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVSymbolSubrecord.java:
```diff
@@ -245,6 +245,88 @@ public String toString() {
+    public static class CVSymbolLData32Record extends CVSymbolData32Record {
+
+        CVSymbolLData32Record(CVDebugInfo cvDebugInfo, String symbolName, int typeIndex, int offset, short segment) {
+            super(cvDebugInfo, CVDebugConstants.S_LDATA32, symbolName, symbolName, typeIndex, offset, segment);
```
This really belongs in the followup PR, but I have added a constructor with a displayname.

stooke commented:
@pejovica thank you for your review, and I'm sorry circumstances forced me to take this long to get back to this PR.  I've incorporated your suggestions.  Should I rebase on head?

pejovica commented:
@stooke No worries, glad you're back!

> Should I rebase on head?

Yes, that would be great.

stooke commented:
> @stooke No worries, glad you're back!
> 
> > Should I rebase on head?
> 
> Yes, that would be great.

@pejovica , I've rebased the PR.  Thank you for your patience.

pejovica commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeSectionBuilder.java:
```diff
@@ -0,0 +1,552 @@
+
+            /* Build all Field List records in field order (FIFO). */
+            for (CVTypeRecord.FieldRecord fieldRecord : fields) {
+                if (FIELD_LIST_RECORD_OVERHEAD + currentFieldList.getEstimatedSize() + CVUtil.align4(fieldRecord.computeSize()) >= CV_TYPE_RECORD_MAX_SIZE) {
```
I'm not sure I understand this condition: the initial value of `CVFieldListRecord.estimatedSize` essentially already takes into account `FIELD_LIST_RECORD_OVERHEAD`, so why add it again here?

Shouldn't we reserve space for `CVIndexRecord` instead? That is, shouldn't the condition be something like this
```suggestion
                if (currentFieldList.getEstimatedSize() + CVUtil.align4(fieldRecord.computeSize()) > CV_TYPE_RECORD_MAX_SIZE - SPACE_FOR_CV_INDEX_RECORD) {
```
where `SPACE_FOR_CV_INDEX_RECORD` is equal to `CVUtil.align4(new CVIndexRecord(0).computeSize())`?

stooke replied:
Done.  I've rearranged your suggestion and added some comments explaining the intent of the calculation.

pejovica commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeRecord.java:
```diff
@@ -235,4 +411,795 @@ public boolean equals(Object obj) {
+
+        static final int INITIAL_CAPACITY = 10;
+
+        /* Size includes type field and record length field. */
+        private int estimatedSize = Short.BYTES * 2;
```
I'd rather keep this as it was. While the value is technically the same in both cases, it's not because the size includes a length field, but because of the alignment. 
```suggestion
        /* Size includes type field but not record length field. */
        private int estimatedSize = CVUtil.align4(Short.BYTES);
```

stooke replied:
Reverted.

pejovica commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVSymbolSubsectionBuilder.java:
```diff
@@ -156,23 +169,35 @@ private void addLineNumberRecords(CompiledMethodEntry compiledEntry) {
+     * @param entry compiled method containing entities whos type records must be added
+     * @return type index of function type
+     */
+    private int addTypeRecords(@SuppressWarnings("unused") CompiledMethodEntry entry) {
```
This is no longer necessary.
```suggestion
    private int addTypeRecords(CompiledMethodEntry entry) {
```

stooke replied:
Done.

pejovica reviewed (approved):
@stooke One last thing, can you please squash everything into one commit before I proceed with the integration.

stooke commented:
> @stooke One last thing, can you please squash everything into one commit before I proceed with the integration.

@pejovica I have pushed the PR as a single commit.

pejovica commented:
Excellent, thanks!
